### ISSUE TITLE
Followup to #368 - Emscripten Docker Image 

### ISSUE THREAD

**Issue by trzecieu** [emscripten-core/emsdk#372]
Hi, I can I'm a part of organization on docker. Thanks.

Looks like there are 2 projects: `emscripten/emscripten-ci` and `emscripten/emsdk` - I thought that we will go under `emscripten/emscripten` name. 

If everything is fine let me know so that I can start building and pushing images 1.38.40... (fastcomp + upstream flavors). 

Here is my small story to complete:
~- [ ] build and publish all current versions~
~- [ ] create dockerhub project description~
~- [ ] update [trzeci/emscripten](https://hub.docker.com/r/trzeci/emscripten/) descriptions to advertise using the official image~
~- [x] hook building image CI (starting with own Jenkins)~
update: changed as https://github.com/emscripten-core/emsdk/issues/372#issuecomment-639986878

I see that hooking building Docker Image with my current infrastructure as a temporary solution. Ideally this should be hooked after the official build is published. 

Issue created for visibility after #368, let me know what you think about plan above and if I have a green light. 
CC: @sbc100, @kripken 

Best regards,
Piotr

**Comment by kripken** [comment]
> Looks like there are 2 projects: emscripten/emscripten-ci and emscripten/emsdk - I thought that we will go under emscripten/emscripten name.

Oh, I thought @sbc100 suggested "emscripten/emsdk" for this image - maybe I misunderstood? I don't have a preference either way, but emsdk does make sense I think.

**Comment by sbc100** [comment]
I did agree that `emscripten/emscripten` was ok, but I'm ok with `emscripten/emsdk` too.  in some say it does make more sense.

**Comment by sbc100** [comment]
Regarding CI, please stick to cricleci so that we only have one CI system.

**Comment by sbc100** [comment]
Otherwise sounds great!  Thanks for working on this.

**Comment by Squareys** [comment]
@sbc100 @trzecieu Hey guys, I highly appreciate your efforts of moving @trzecieu's Dockerfiles to an official image!
What's the status on this?

**Comment by sbc100** [comment]
I think the status is that we'd love to see it move forward but its been stalled for a few months.  If yourself or @trzecieu wants to get this going agin it would be more than welcome!

**Comment by sbc100** [comment]
Also, lets drop that fastcomp image in just ship llvm upstream.

**Comment by trzecieu** [comment]
Hi @Squareys, @sbc100 : Thank you for poking and being interested in help, and in general what's the status. I appreciate this. 
I had to put on hold all hobbyist projects due family reasons for a while, just started doing some work last week and this is something that I'd like to finalize.  

What does it mean in practice? This week I'll investigate why some last build are failing on my CI and process collected issues. 
And actually help is appreciated, especially in the way how should we hook building docker image to current CI of emscripten-core. 

I'm open for cooperation and feedback.

**Comment by Squareys** [comment]
@trzecieu @sbc100 I'd be happy to volunteer for the setup on Circle CI, but I am unsure of how the process is here.
I guess I can start off by creating the Circle CI config for a fork and then contribute the dotfile through a PR to this repo?

**Comment by Squareys** [comment]
Since the Circle CI config changes have been merged, what's left to do is to:

 - [ ] Set CircleCI `DOCKER_PASS` and `DOCKER_USER` on the emsdk Circle CI account (@haraldreingruber [mentioned](https://github.com/emscripten-core/emsdk/pull/513#discussion_r435738602) @kripken might have access to the credentials)
 - [ ] Build and publish existing tags, as @trzecieu mentioned. He would need to be added to the emscripten/emsdk docker hub (His docker user is [trzeci](https://hub.docker.com/u/trzeci)].

**Comment by trzecieu** [comment]
About point nr 2, an interesting part is that I'm a member of the `emscripten` org, team `emsdk` - maybe there is some missing settings allowing me to push.

**Comment by kripken** [comment]
About point 1, what exactly should I do? (bear in mind I am not familiar with docker... what values should those fields have, to start with?)

About point 2, I see the emscripten/emsdk repo, but am not sure how to give @trzecieu additional permissions there. I don't see an option to add a user as a collaborator, and in the permissions tab I can make an "emsdk" team with read+write permissions but I don't see how to add people to that team...

**Comment by trzecieu** [comment]
Hi! 
About point 2, Just tested and it looks I can push image to `emscripten/emsdk` That's great! I will check on pushing couple of images. 
About point 1, It's possible to generate an access token (https://hub.docker.com/settings/security) for this purpose that can be used instead of password on CircleCI - also it's easy to revoke if something goes wrong.

**Comment by trzecieu** [comment]
Works like a charm, awesome work @Squareys 

@kripken on your end it's needed to add `DOCKER_PASS` (from generated token) and matching `DOCKER_USER` that has a right to push to `emscripten/emsdk` that's all and starting from next release it will build the image automatically

**Comment by sbc100** [comment]
How/where are we supposed to set DOCKER_PASS+DOCKER_USER?

**Comment by trzecieu** [comment]
Under https://app.circleci.com/settings/project/github/emscripten-core/emsdk/environment-variables

**Comment by trzecieu** [comment]
Tags 1.39.0 ... 1.39.17 are in place: https://hub.docker.com/repository/docker/emscripten/emsdk/tags

**Comment by Squareys** [comment]
> pushed images as tags like 1.39.0 instead of 1.39.0-upstream

I think @sbc100 wanted you to drop the upstream anyway, since there is _only_ -upstream images anyway?

**Comment by haraldreingruber** [comment]
> @kripken on your end it's needed to add `DOCKER_PASS` (generated token) and matching `DOCKER_USER` that has a right to push to DockerHub `emscripten/emsdk` that's all and starting from next release it will build the image automatically.

In case you want to create the access token from another user, @kripken I guess you could have a look at how the push rights were configured for me on https://hub.docker.com/repository/docker/emscripten/emscripten-ci and do the same for the emsdk project. Just an idea.

**Comment by trzecieu** [comment]
OK, let's do this: https://github.com/emscripten-core/emsdk/pull/519

**Comment by trzecieu** [comment]
Is it possible to get access to logs of CircleCI and see why the image wasn't released?

**Comment by sbc100** [comment]
If you look at the commit history of master you can see the tests that ran against the latest commit.

I think the logs are all public.  Here is the link I got to:

https://app.circleci.com/pipelines/github/emscripten-core/emsdk/445/workflows/e22e88b1-e568-4376-9209-e43bf10a1343/jobs/1628/steps

**Comment by trzecieu** [comment]
Thanks! So it looks like tag was compiled correctly but it has failed on docker login command

**Comment by sbc100** [comment]
I guess the environment variables have not yet been added.

**Comment by sbc100** [comment]
@kripken do you what to try to add those environmet one?)nt variables?   Or would you like me to?

**Comment by sbc100** [comment]
I've added my DOCKER_USER and DOCKER_PASS to circleci and successfully published a tagged version of emsdk at https://hub.docker.com/r/emscripten/emsdk/tags.

Now each tagged release should automatically be published going forward.  

@trzecieu for the next steps would you mind copying the description text over from your old docker hub account and perhaps marking that one as old or somehow redirecting people?

**Comment by trzecieu** [comment]
Absolutely! I think we can do something like I had there, that is on every docker push, description from Readme.md will be send to dockerhub. So small automation. 

I will need to think how to redirect people in the best way, description somehow started to work.

**Comment by Squareys** [comment]
@sbc100 @trzecieu Amazing work you guys, thanks for getting this out there!

**Comment by RReverser** [comment]
Just to be clear - is the migration here finished and automated on emsdk side? If so, could we now document Docker images as officially supported on https://emscripten.org/docs/getting_started/downloads.html?

**Comment by sbc100** [comment]
Yup, I think this is almost all done now.  We publish automatically via github actions to docker hub now.

We still need a README there and it would be nice to redirect people from trzecieu old dockerhub listing, and we should update the emscripten docs as you say... so we can leave this issue open until we've wrapped up those loose ends.

**Comment by RReverser** [comment]
Not sure if it's just a local issue but running into something new after switching Docker to `emscripten/emsdk`:

```
error: failure to process js library "library_sockfs.js": Missing C define SOCK_CLOEXEC! If you just added it to struct_info.json, you need to ./emcc --clear-cache,
error: use -s VERBOSE to see more details
Internal compiler error in src/compiler.js!
Please create a bug report at https://github.com/emscripten-core/emscripten/issues/
with a log of the build and the input files used to run. Exception message: "Missing C define SOCK_CLOEXEC! If you just added it to struct_info.json, you need to ./emcc --clear-cache" | undefined
```

**Comment by sbc100** [comment]
Looks like a real bug, not sure how it could be Docker specific..   IIRC the reference to SOCK_CLOEXEC in the library was recently added.

**Comment by sbc100** [comment]
Do you have an easy command to repro that error?

**Comment by RReverser** [comment]
It failed for me on autoconf step in MozJPEG (looks like failing on a simple `int main() { }` test), I can try to provide more details tomorrow.

**Comment by RReverser** [comment]
Can't reproduce in isolation so far, so maybe some sort of cache that got shared from previous image. I'll try to remove all folders used for build and restart anew.

**Comment by RReverser** [comment]
Hmm, no, that still fails. A bit more from `config.log`:

```
configure:3708: checking whether the C compiler works
configure:3730: /emsdk/upstream/emscripten/emcc -Os -flto --llvm-lto 1 -Os -flto --llvm-lto 1 -Os -flto --llvm-lto 1 conftest.c  >&5
emcc:WARNING: --llvm-lto ignored when using llvm backend
emcc:WARNING: --llvm-lto ignored when using llvm backend
emcc:WARNING: --llvm-lto ignored when using llvm backend
error: failure to process js library "library_sockfs.js": Missing C define SOCK_CLOEXEC! If you just added it to struct_info.json, you need to ./emcc --clear-cache,
error: use -s VERBOSE to see more details
Internal compiler error in src/compiler.js!
Please create a bug report at https://github.com/emscripten-core/emscripten/issues/
with a log of the build and the input files used to run. Exception message: "Missing C define SOCK_CLOEXEC! If you just added it to struct_info.json, you need to ./emcc --clear-cache" | undefined
emcc: error: '/emsdk/node/12.18.1_64bit/bin/node /emsdk/upstream/emscripten/src/compiler.js /tmp/tmp7ayf3r33.txt' failed (1)
configure:3734: $? = 1
configure:3772: result: no
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "mozjpeg"
| #define PACKAGE_TARNAME "mozjpeg"
| #define PACKAGE_VERSION "3.3.1"
| #define PACKAGE_STRING "mozjpeg 3.3.1"
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| #define PACKAGE "mozjpeg"
| #define VERSION "3.3.1"
| /* end confdefs.h.  */
| 
| int
| main ()
| {
| 
|   ;
|   return 0;
| }
configure:3777: error: in `/src/node_modules/mozjpeg':
configure:3779: error: C compiler cannot create executables
```

**Comment by RReverser** [comment]
I don't understand what's special about it being run by `./configure` vs me running `emcc` from the same Docker image with same params and code...

**Comment by RReverser** [comment]
Ah finally found where these Emscripten files were cached and managed to build.

I wonder why Emscripten doesn't automatically invalidate the cache on version mismatch in this case. Is it worth raising as a separate issue?

**Comment by sbc100** [comment]
I don't don't understand what is going wrong here.  The struct_info.json that ships with the emsdk embedded in the cache should always be up-to-date.  

Is this a general emsdk issue or specific to the docker image?   

Do you have a way I can repro this issue?

**Comment by RReverser** [comment]
So the problem here was using custom EM_CACHE location to make sure that cache survives outside the Docker and libraries don't get rebuilt every time... but, of course, that causes problems on Emscripten upgrade like one I had above.

Which makes me wonder whether such outdated cache could be detected earlier in the pipeline, and, instead of leading to compiler errors on upgrade, simply invalidated and rebuilt.

**Comment by sbc100** [comment]
Ah yes that sounds like a bug.

The current approach is that we write a file called `.emscripten_sanity` alongside the config file.    It contains the emscripten version and hash of the config file.    If this changes (i.e. if you upgrade emscripten or edit the config in any way) it will clear the cache.   In this case emscripten and the config and the sanity file were all updated in sync do nothing was cleared :(

One possible solution would be to write sanity file instead inside the cache itself.  I  think that would support using an external cache like this.   I think this is indeed a separate issue that applies the emsdk in general (and anyone else who packages a config and santiy file together and ships both).

Regarding the idea of using the external / persistent cache:  One issue is that if you do this you want to make sure you take advantage of the pre-built libraries in emsdk.  We now ship all of the libraries and all the ports pre-built so you would want to at least copy of the contents of the cache that we ship.

Assuming users do want to write to the cache and have those chance persist what do you think is the best way to do this?  Would saving a docker image layer make sense?   That way you wouldn't need to set EM_CACHE at all right?  I'm assuming you are moving the cache outside of the docker image completely?  Is that easier/better than mutating/adding to the docker image?

**Comment by RReverser** [comment]
> The current approach is that we write a file called `.emscripten_sanity` alongside the config file.

What's a config file? I don't think we use that.

> One possible solution would be to write sanity file instead inside the cache itself. I think that would support using an external cache like this.

That sounds like reasonable idea, yeah. I think even putting Emscripten version inside `generated_struct_info.json` could help (right now it seems versionless).

> Regarding the idea of using the external / persistent cache: One issue is that if you do this you want to make sure you take advantage of the pre-built libraries in emsdk.

Sure, in most cases that's enough, but sometimes, as a user, you want custom set of options that standard build doesn't provide. E.g. before the primary reason for adding such an external cache was https://github.com/trzecieu/emscripten-docker/issues/67 where LTO builds of the library weren't prebuilt as part of the image, and instead were getting built on each run.

But there are other options that can cause similar behaviour - between different optimisation / debug levels or added sanitizers, there is always some edge case that might require custom builds and for those cases it might be worth to support them better.

> Assuming users do want to write to the cache and have those chance persist what do you think is the best way to do this? Would saving a docker image layer make sense?

Yeah a separate Docker image layer would solve this too, although you'd still need to trigger build without actual source files but with required version + params first.

I agree it's a better way to solve Emscripten image upgrades with Docker, just not something we do yet - historically we have a single `build.sh` that is getting executed in context of Docker container.

From what you're saying, it still sounds like this issue is worth solving at least for non-Docker cases though?

**Comment by sbc100** [comment]
> > The current approach is that we write a file called `.emscripten_sanity` alongside the config file.
> 
> What's a config file? I don't think we use that.

emscripten always needs a `.emscripten` config file and writes a `.emscripten_sanity` alongside it on first use.    With emsdk and with the docker image it all comes pre-packaged so you don't need to worry about (at least thats that idea).  

emscripten will nuke the entire cache if you change the version number or the config file.

**Comment by RReverser** [comment]
Ah, right, so in this case sounds like the problem is that `.emscripten` + `.emscripten_sanity` are prepackaged, but cache external, so they're not validated against each other.

**Comment by RReverser** [comment]
@sbc100 Thanks for creating an issue!

**Comment by sbc100** [comment]
Yes.. the assumption was that the cache is *also* pre-packaged in which case all is good... but I guess we forgot about those pesky environment variables :)

**Comment by RReverser** [comment]
https://xkcd.com/1172/ as usual :)

**Comment by sbc100** [comment]
My god that xkcd is so appropriate to VM and toolchain development.

Reminds me of this issue from yesterday: https://github.com/emscripten-core/emscripten/pull/11586.   I would never have imagined it would be useful to pre-process multiple C files in a single command, but separately.   Mostly because (a) this only works with output to stdout (cannot use -o in this case) (b) the result will almost certainly not compile because all the includes will be duplicated.    But in this case they don't care about compiling that result.. and gcc and clang can both do this for some reason.

**Comment by RReverser** [comment]
> Absolutely! I think we can do something like I had there, that is on every docker push, description from Readme.md will be send to dockerhub. So small automation.
> 
> I will need to think how to redirect people in the best way, description somehow started to work.

Now that the side issue is fixed, coming back to the original context - @trzecieu is this something on your radar?

**Comment by sbc100** [comment]
Can this be closed now?

**Comment by RReverser** [comment]
I believe so.
